File: Key Project/src/org/mathiasGarnier/keyProject/object/Key.java
package org.mathiasGarnier.keyProject.object;

import org.mathiasGarnier.keyProject.enumeration.EnumPlace;
import org.mathiasGarnier.keyProject.utils.Utils;

public class Key implements Cloneable {

    private String keyOwner;
    private int actionRange; //Well if a door is to 3 actionRange, a key with 2 actionRange can't be open the door
    private EnumPlace keyFor; //Well the "house", the "garden"...
    
    public Key(String nKeyOwner, int nActionRange, EnumPlace nKeyFor) {
	
	this.setKeyOwner(nKeyOwner);
	this.setActionRange(nActionRange);
	this.setKeyFor(nKeyFor);
    }
    
    //For the actionRange : get a Random number, and test if the random number is > || < to know if actionRange is adequate ^^'. -> 2D
    
    /*
     * FUNCTION
     */
    
    /**
     * @param door a {@link Boolean} for the door state
     * @return a {@link Boolean} with the state of the dooor
     */
    public boolean isDoorOpen(EnumPlace door) {
	
	if (door.isEnumPlaceState()) return true;
	else return false;
    }
    
    /**
     * @param door an {@link EnumPlace} for the door state
     */
    public void openDoor(EnumPlace door) {
	
	if (!(door.isEnumPlaceState())) return;
		
	else door.setState(true);
    }
    
    /**
     * @param door an {@link EnumPlace} for the door state 
     * @return a {@link Boolean} true if door can be opened, false if door can't be opened
     */
    public boolean canOpen(EnumPlace door) {
	
	if (door.equals(getKeyFor())) return true;
	else return false;
    }
    
    /**
     * @return a {@link Boolean} true if door is reachable, false  if door isn't reachable
     */
    public boolean reachableDoor(){
	
	Utils u = new Utils();
	int nU = u.getRandomNumber(10); //CHANGE THIS LATER
	
	if (nU > getActionRange()) return false;
	else return true;
    }
    
    /*
     * GETTER AND SETTER
     */
    
    /**
     * @return the name
     */
    public String getKeyOwner() {
	
	return keyOwner;
    }

    /**
     * @param name the name to set
     */
    public void setKeyOwner(String name) {
	
	this.keyOwner = name;
    }

    /**
     * @return the actionRange
     */
    public int getActionRange() {
	
	return actionRange;
    }

    /**
     * @param actionRange the actionRange to set
     */
    public void setActionRange(int actionRange) {
	
	this.actionRange = actionRange;
    }

    /**
     * @return the keyFor
     */
    public EnumPlace getKeyFor() {
	
	return keyFor;
    }

    /**
     * @param keyFor the keyFor to set
     */
    public void setKeyFor(EnumPlace keyFor) {
	
	this.keyFor = keyFor;
    }
    
    /*
     * SUPER FUNCTION
     */
    
    /**
     * @param o the compare {@link Object}
     * @return a {@link Boolean} true if key object is equals to an o {@link Object}, false if any key object is different to an o {@link Object}
     */
    public boolean equals(Object o) {
	
	if(!(o instanceof Key)) return false;
	
	Key k = (Key) o;
	
	return keyOwner.equals(k.keyOwner) && actionRange == k.actionRange && keyFor.equals(k.keyFor);
    }
    
    /**
     * 
     * @param o the compare {@link Object}
     * @return a {@link Boolean} true if key {@link Object} is equals to an o {@link Object}, false if any Key {@link Object} is different to an o {@link Object}
     */
    public boolean equalsWithoutKeyFor(Object o) {
	
	if (!(o instanceof Key)) return false;
	
	Key k = (Key) o;
	
	return keyOwner.equals(k.keyOwner) && actionRange == k.actionRange;
    }
}

File: cKeyProject/Object/Key.cpp
//
// Created by Yapix on 25/02/2017.
//

#include "Key.h"

/* CONSTRUCTOR */

/**
 * @param nKeyOwner The owner of this key
 * @param nActionRange The action range of the key (Well if a door is set to actionRange = 3,
 * 					   a key set to actionRange = 2 can't open the door.
 * @param nKeyFor
 * @usage Construct a new Key in function of his owner and his action range
 * @example Key k("jean louis", 10);
 */
Key::Key(std::string nKeyOwner, int nActionRange, Place nKeyFor) : keyOwner(nKeyOwner), actionRange(nActionRange), keyFor(nKeyFor){}

/* GETTERS AND SETTERS */

/**
 * @usage Get the owner's name
 * @example Key k("jean louis", 10);
 * 			std::cout << k.getKeyOwner(); //Return "jean louis"
 */
std::string Key::getKeyOwner() {

	return keyOwner;
}

/**
 * @param pOwner New value of Owner of the key
 * @usage Set the new value of Owner of the key
 * @example Key k("jean louis", 2);
 * 			k.setKeyOwner("jean");
 * 			k.getKeyOwner(); //Return "jean"
 */
void Key::setKeyOwner(std::string pOwner) {

	this->keyOwner = pOwner;
}

/**
 * @usage Get the action range of the Key
 * @example Key k("jean louis", 3);
 * 			std::cout << k.getActionRange(); //Return 3
 */
int Key::getActionRange() {

	return actionRange;
}

/**
 * @param pActionRange New value of actionRange
 * @usage Set the new value of Key's actionRange
 * @example Key k("jean louis", 2);
 * 			k.setActionRange(k.getActionRange() * 3);
 * 			k.getActionRange(); //Return 6
 */
void Key::setActionRange(int pActionRange) {

	this->actionRange = pActionRange;
}

Place Key::getKeyFor() {

	return keyFor;
}

void Key::setKeyFor(Place pKeyFor) {

	this->keyFor = pKeyFor;
}

/* UTILITY FUNCTION */

/**
 * @param other This is the compared parameter (to this -> this.equals(other))
 * @usage Check if this and other are equals
 * @example Key k("jean louis", 10), k2("jean louis", 10);
 * 			std::cout << std::boolalpha << (std::is_same<decltype(k), decltype(k2)>::value) << std::endl;
 * 			std::cout << std::boolalpha << k.equals(k2); // Return true
 */
bool Key::equals(Key other) {

	return (getKeyOwner() == other.getKeyOwner() && getActionRange() == other.getActionRange());
}
